### ISSUE TITLE
serial out locks out

### ISSUE THREAD

**Issue by pelikhan** [microsoft/pxt-microbit#3499]
It seems that the tx buffer of the serial fills up and serial.write locks up.

This simple program does not lock in v1; but eventually locks up after 9 iterations on v2.
```
basic.forever(function () {
    led.toggle(1, 0)
    serial.writeLine("aaaa")
    basic.pause(500)
})
basic.forever(function () {
    led.toggle(0, 0)
    basic.pause(50)
})
```

**Comment by finneyj** [comment]
Thanks @pelikhan. Odd.

**Comment by finneyj** [comment]
I can't repro this precisely. This works fine against current head of master... I'll keep digging - it probably is codal side.

```
#include "MicroBit.h"
#include "Tests.h"

MicroBit uBit;

void
task2()
{
    while(1)
    {
        uBit.display.image.setPixelValue(0,0,255);
        uBit.sleep(50);

        uBit.display.image.setPixelValue(0,0,0);
        uBit.sleep(50);
    }
}

void
task1()
{
    while(1)
    {
        uBit.display.image.setPixelValue(1,0,255);
        uBit.serial.send("aaaa");
        uBit.sleep(500);

        uBit.display.image.setPixelValue(1,0,0);
        uBit.serial.send("aaaa");
        uBit.sleep(500);
    }
}

int 
main()
{
    uBit.init();

    create_fiber(task2);
    create_fiber(task1);

    while(1)
        uBit.sleep(10000);
}
```

**Comment by finneyj** [comment]
Just for tracking - This issue seems to be a problem codal side brought on by the additional padding that MakeCode injects in Serial.ts to align sends to a 32 byte boundary.  The following does repro, which is a closer approximation to what MakeCode is doing under the hood:

```
#include "MicroBit.h"
#include "Tests.h"

MicroBit uBit;

int writeLinePadding = 32;

void
task2()
{
    while(1)
    {
        uBit.display.image.setPixelValue(0,0,255);
        uBit.sleep(50);

        uBit.display.image.setPixelValue(0,0,0);
        uBit.sleep(50);
    }
}

void tx(ManagedString text)
{
    uBit.serial.send(text);

    if (writeLinePadding > 0) {
        int r = (writeLinePadding - (text.length() + 1) % writeLinePadding) % writeLinePadding;
        for (int i = 0; i < r; ++i)
            uBit.serial.send(" ");
    }
    uBit.serial.send("\n");
}

void
task1()
{
    while(1)
    {
        uBit.display.image.setPixelValue(1,0,255);
        tx("aaaa");
        uBit.sleep(500);

        uBit.display.image.setPixelValue(1,0,0);
        tx("aaaa");
        uBit.sleep(500);
    }
}

int 
main()
{
    uBit.init();

    //uBit.audio.setSpeakerEnabled(false);
    //uBit.audio.soundExpressions.play("happy");
    //uBit.audio.virtualOutputPin.setAnalogValue(512);
    //uBit.audio.virtualOutputPin.setAnalogPeriodUs(2272);

    create_fiber(task2);
    create_fiber(task1);

    while(1)
        uBit.sleep(10000);
}
```

**Comment by finneyj** [comment]
Found a small race condition in Serial.cpp that seemed to be causing this behaviour reproducibly in the C++ sample above. 

(Hopefully) Fixed in v0.2.13

**Comment by pelikhan** [comment]
Looks fixed. I'll let this run overnight later.
